elements/playerbufftimers.lua

```diff
@@ -0,0 +1,290 @@
+    pbt:SetPoint('BOTTOM', self, 'TOP')
+    pbt:SetSize(230, 20 * 4)
+
+    -- Register with oUF
+    self.PlayerBuffTimers = pbt
+--]]
+
+local _, ns = ...
+local oUF = ns.oUF or oUF
+
+local growth_direction = {
+	TOPLEFT     = { 1, -1},
+	TOPRIGHT    = {-1, -1},
+	BOTTOMLEFT  = { 1,  1},
+	BOTTOMRIGHT = {-1,  1},
+}
+
+local function SetPosition(element)
+	local width = (element.width or element:GetWidth()) + (element['spacing-x'] or element.spacing or 0)
+	local height = (element.height or 16) + (element['spacing-y'] or element.spacing or 0)
+	local anchor = element.anchor or 'TOPLEFT'
+	local direction = growth_direction[anchor]
+	if(not direction) then
+		local anchors
+		for key in next, growth_direction do
+			if(not anchors) then
+				anchors = key
+			else
+				anchors = string.format('%s, %s', anchors, key)
+			end
+		end
+		error(string.format('Invalid anchor %s. Should be one of %s', anchor, anchors))
+	end
+	local x = direction[1]
+	local y = direction[2]
+	local cols = math.floor(element:GetWidth() / width + .5)
+	local rows = math.floor(element:GetHeight() / height + .5)
+
+	local hidden = 0
+	for i = 1, #element do
+		local timer = element[i]
+		if(timer:IsShown()) then
+			local col, row
+			if(element.primaryAxis == 'x') then
+				col = (i - 1 - hidden) % cols
+				row = math.floor((i - 1 - hidden) / cols)
+			else
+				col = math.floor((i - 1 - hidden) / rows)
+				row = (i - 1 - hidden) % rows
+			end
+			timer:ClearAllPoints()
+			timer:SetPoint(anchor, element, anchor, col * width * x, row * height * y)
+		else
+			hidden = hidden + 1
+		end
+	end
+end
+
+local function onEnter(timer)
+	if(not timer:IsVisible()) then return end
+
+	if(timer.Time) then
+		timer.Time:Show()
+	end
+
+	GameTooltip:SetOwner(timer, 'ANCHOR_BOTTOMRIGHT')
+	GameTooltip:SetText(timer.powerName, 1, 1, 1)
+	GameTooltip:AddLine(timer.powerTooltip, nil, nil, nil, true)
+	GameTooltip:Show()
+end
+
+local function onLeave(timer)
+	if(timer.Time) then
+		timer.Time:Hide()
+	end
+	GameTooltip:Hide()
+end
+
+local function onUpdate(timer)
+	local timeLeft = timer.expiration - GetTime()
+
+	if(timeLeft > 0) then
+		timer:SetValue(timeLeft)
+
+		if(timer.Time and timer.Time:IsVisible()) then
+			--[[ Callback: timer:CustomTime(timeLeft)
+			Called after the timer's remaining time changed.
+
+			* self     - the timer widget
+			* timeLeft - the remaining timer duration in seconds
+			--]]
+			if(timer.CustomTime) then
+				timer:CustomTime(timeLeft)
+			else
+				timer.Time:SetFormattedText('%d / %d', timeLeft, timer.duration)
+			end
+		end
+	else
+		timer:Hide()
+		--SetPosition(timer:GetParent()) -- TODO: remove after testing
+	end
+end
+
+local function UpdateTimer(timer, duration, expiration, barID, auraID)
+	timer:SetMinMaxValues(0, duration)
+	timer.duration = duration
+	timer.expiration = expiration
+	timer.auraID = auraID
+end
+
+local function CreateTimer(element, index)
+	local timer = CreateFrame('StatusBar', nil, element)
```

Comment:
Auras are in charge of a lot of extra widgets, this one only has one, and is very comparable to elements like ClassPower, Totems and Runes. Runes and this one is pretty much identical apart from the updating, yet it lets the layout create the bars.